File: slash/dd/color.ts
import { AttachmentBuilder, ChatInputCommandInteraction, EmbedBuilder, SlashCommandSubcommandBuilder } from 'discord.js';
import Bot from '../../bot/bot';
import SlashCommand from '../../bot/slashcommand';
import fetch from 'node-fetch';

export default class Color extends SlashCommand {
  constructor(client?: Bot) {
    const data = new SlashCommandSubcommandBuilder()
      .setName('color')
      .setDescription('Emulate the forge color mechanic')
      .addStringOption(option => option
        .setName('color-string')
        .setDescription('The color string (the text between <color: and >)')
        .setRequired(true)
      );
    super(data, client, { belongsTo: 'dd' });
  }

  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const colorString = interaction.options.getString('color-string');
    if (!/^[,\s]*-?\d{1,3}([,\s]*-?\d{1,3}){0,2}[,\s]*$/.test(colorString)) {
      const embed = new EmbedBuilder()
        .setColor(this.client.config.colors.RED)
        .setTitle('Invalid format')
        .setDescription('Take your character name, remove `<color:` and `>` so you\'re only entering the numbers and commas.');
      interaction.reply({ embeds: [embed] });
      return;
    }

    const vals = colorString.match(/-?\d{1,3}/g).map(val => parseInt(val));

    for (let i = 0; i < vals.length; i++) {
      while (vals[i] < 0) vals[i] += 255;
      while (vals[i] > 255) vals[i] -= 255;
    }

    const hexifier = (x: number) => x.toString(16).padStart(2, '0');
    const hex = hexifier(vals[0]) + hexifier(vals[1] ?? 255) + hexifier(vals[2] ?? 255);
    const apiKey = `https://singlecolorimage.com/get/${hex}/200x200`;

    const response = await fetch(apiKey);
    const buffer = await response.buffer();
    const attachment = new AttachmentBuilder(buffer, { name: 'color.png' });

    const embed = new EmbedBuilder()
      .setColor(`#${hex}`)
      .setTitle('Resulting color')
      .setDescription(`\`<color:${vals.join(',')}>\` is equivalent to \`<color:${vals[0]},${vals[1] ?? 255},${vals[2] ?? 255}>\``)
      .setImage('attachment://color.png');
    interaction.reply({ embeds: [embed], files: [attachment] });
  }
}

File: src/commands/dd/color.rs
use regex::{Match, Regex};
use serenity::all::{CommandInteraction, CommandOptionType};
use serenity::async_trait;
use serenity::builder::{
  CreateAttachment,
  CreateCommandOption,
  CreateEmbed,
  CreateEmbedFooter,
  CreateInteractionResponseMessage
};
use serenity::client::Context;

use crate::color::Colors;
use crate::commands::SlashSubCommand;
use crate::database::Database;
use crate::interaction::{BetterResponse, InteractionCustomGet};
use crate::Result;

#[derive(Default)]
pub struct Color;

#[async_trait]
impl SlashSubCommand for Color {
  fn register(&self) -> CreateCommandOption {
    CreateCommandOption::new(
      CommandOptionType::SubCommand,
      "color",
      "Emulate the forge color mechanic"
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::String,
        "color-string",
        "The color string (the text between <color: and >)"
      )
      .required(true)
    )
  }

  async fn execute(&self, ctx: &Context, interaction: &CommandInteraction, _: &Database) -> Result<()> {
    let color_string = interaction.get_string("color-string").unwrap();

    let re = Regex::new(
      r"^[,\s]*(?<red>-?\d{1,3})([,\s]*,[,\s]*(?<green>-?\d{1,3}))?([,\s]*,[,\s]*(?<blue>-?\d{1,3}))?[,\s]*$"
    )
    .unwrap();

    if !re.is_match(&color_string) {
      let embed = CreateEmbed::new()
        .color(Colors::Red)
        .title("Invalid format")
        .description(
          "Take your character name, remove `<color:` and `>` so you're only entering the numbers and commas."
        );

      interaction.reply_embed(ctx, embed).await?;
      return Ok(());
    }

    let caps = re.captures(&color_string).unwrap();
    let red = caps.name("red").map(|m| into_constrained(m, 256)).unwrap();
    let green = caps.name("green").map(|cap| into_constrained(cap, 256)).unwrap_or(255);
    let blue = caps.name("blue").map(|cap| into_constrained(cap, 256)).unwrap_or(255);

    let hex = hexify(red, green, blue);
    let url = format!("https://singlecolorimage.com/get/{hex:0>6x}/300x200");
    let image_bytes = reqwest::get(url).await?.bytes().await?;

    let image = CreateAttachment::bytes(image_bytes, "color.png");

    let embed = CreateEmbed::new()
      .color(hex)
      .title("Resulting color")
      .description(format!(
        "`<color:{}>` is equivalent to `<color:{red},{green},{blue}>`",
        caps.get(0).unwrap().as_str()
      ))
      .attachment("color.png")
      .footer(CreateEmbedFooter::new(format!("Hex: {hex:0>6x}")));

    let data = CreateInteractionResponseMessage::new().embed(embed).add_file(image);
    interaction.reply(ctx, data).await?;
    Ok(())
  }
}

fn hexify(r: i32, g: i32, b: i32) -> u32 {
  ((r as u32) << 16) + ((g as u32) << 8) + (b as u32)
}

/// Converts a capture into a number beween `0` (incl) and `n` (excl)
fn into_constrained(capture: Match<'_>, n: i32) -> i32 {
  modulo(into_number(capture), n)
}

/// Converts a capture into an `i32`
fn into_number(capture: Match<'_>) -> i32 {
  capture.as_str().parse::<i32>().unwrap()
}

/// A number `i` modulo another number `n` is always in the range `[0..n]`
fn modulo(i: i32, n: i32) -> i32 {
  i.rem_euclid(n)
}
